### DISCUSSION TITLE
M6$lv.X

### DISCUSSION THREAD

**Discussion by AlainZuur** [JenniNiku/gllvm#170] [General]
Am I correct that in the development version, M6$lv.X is not a matrix anymore? I think in the CRAN version it was. No big deal...I just add: 

X <- as.matrix(M6$lv.X)

Alain

**Comment by BertvanderVeen** [comment]
Great stuff Alain! Also related to one of my recent commits.

**Comment by BertvanderVeen** [comment]
Woops, closed this as I mistook it for the bug report in #166.

For the emmeans function I wrote in the discussion I needed both the original matrix and the design matrix for the predictors in the LVs. So, model$lv.X is now the original data.frame passed to the function, and model$lv.X.design now holds the design matrix instead. It's cleaner to separate things that way, and corresponds to the organisation of matrices for "main" X.

I'll make a note in "news" of this change.

**Reply by AlainZuur** [reply]
I think you made a (crucial) spelling mistake? Just to clarify...model$lv.X  is now (you wrote no*t*) the X matrix with original data.

Alain

**Reply by BertvanderVeen** [reply]
Yes, that should be "now". I've put some code into place to ensure backward compatibility with "old" models. Should probably have done that the other day anyway.

**Reply by AlainZuur** [reply]
From somewhere of your website, or maybe from the R code that came with your 2023 paper, you provided some code with the name partR2.gllvm. It uses object$lv.X. That may be be affected by this change.

I will hard-code it via the model.matrix.

Ah..it is from here:
https://zenodo.org/records/7284339

Alain

**Reply by BertvanderVeen** [reply]
Yes, I already pushed a new version to the zenodo repo this morning that takes the change into account.

**Reply by BertvanderVeen** [reply]
Note that I also discovered a bug in that function; the used degrees of freedom did not reflect the constraints in the canonical coefficients. This might have a minor effect on the statistic; for 2LVs 1 constraint is added (1df removed) so that the variation explained might go down *slightly*.

**Reply by AlainZuur** [reply]
That is version 5 I assume. You are much deeper into this than me...and I have no clue what it does if someone includes a categorical covariate with > 2 levels:

But in that case, should 
K <- ncol(object$lv.X)

not be 

K <- ncol(object$lv.X.design)

It is looping over the columns of object$lv.X. So, it will try to get an R2 for a factor. But K is the number of covariates...not the number of parameters. But maybe I am talking nonsense here?


Alain

**Reply by BertvanderVeen** [reply]
In the new version it should be, yes. Hence for compatibility it includes the following line before the one you quoted:

`
 if(!is.null(object$lv.X.design))object$lv.X <- object$lv.X.design
`

I.e., in the new version, lv.X gets replaced with the design matrix, and in the old version it is still the design matrix.

**Reply by AlainZuur** [reply]
Ah..I didn't see that.
Thanks!
